README.md
# BCS.client

**PLEASE NOTE!** The current version of this library requires BCS firmware 4.0.0 or higher.

A node.js implementation of a data client (consumer) for the [BCS-460 and BCS-462][ecc] brewery automation controllers. BCS.client uses the [API][api-docs] native to BCS devices to communicate with them over http.

## Purpose

Long-term process logging is a primary goal for this project. This library makes that easy to do, and easy to automate. This project does not attempt to replace any of the functionality of the BCS UI, nor does it provide logging in of itself. Instead, it aims to enable inter-operability with other projects, processes, and devices by providing a robust consistent interface.

In addition to providing a simple, easy to use communications library, read-through caching has been implemented and thoroughly battle tested. This ensures that any amount of load applied to the library can be tolerated, without saturating the BCS device with HTTP requests.

#### Advantages vs. Direct API Calls

* Easily integrate real-time BCS data into any node.js application on a reachable network.
* Automatic read-through caching engine drastically reduces BCS device load while supporting theoretically unlimited simultaneous requests.
* Rate limiting ensures that even structures which must be refreshed often (like temperatures, for example) don't cause undue BCS device load.
* Wraps all device communications in node-standard, error-first callback patterns.

## Quick Start

Include the `bcs.client` module in your `package.json`, or manually install with `npm install bcs.client`.

````javascript
var Device = require('bcs.client');

var device = new Device('192.168.1.1', function (e, info) {
	
	// device info after setup
	console.log(info);
	/*
		{
		  "type": "BCS-460",
		  "version": "4.0.0",
		  "build": "c5bda61",
		  "name": "Seeker Brewing Co."
		}
	*/
	
	// get the current details of temp probe 2
	device.read('temp/2', function (e, probe) {
		console.log(probe);
		/*
			{
			  "name": "Conical B",
			  "temp": 665,
			  "setpoint": 690,
			  "resistance": 12969,
			  "enabled": true,
			  "coefficients": [
			    0.0011371546,
			    0.0002325949,
			    9.5299999999e-8
			  ]
			}
		*/
	});
	
});
````
## Methods

### new Device(hostname, [port], callback)

Returns a new `Device` instance, connected to the BCS device at the given address.

* `hostname` - ip address or host name of device
* `port` - port of device [80]
* `callback(err, info)`
	* `err` - *Error instance*, or *null/undefined*
	* `info` - device info object

### device.read(target, callback)

* `target` - endpoint you wish to read, ex. *temp/0*
* `callback(err, json)`
	* `err` - *Error instance*, or *null/undefined*
	* `json` - the endpoint json from the device

## Load Management

You can hit `device.read()` as often as you want, as hard as you want. The read-through cache will make sure the device only has to answer as many questions as absolutely needed. Here's the cache stats after a test example with 60,000 simultaneous requests, spread across 6 of the device's API read endpoints:

````javascript
// read-through cache statistics, 60k simultanous requests to 6 endpoints.
// that's a lot of bytes the device didn't need to handle!
{ hits: 59994,
  misses: 6,
  requests: 6,
  responses: 6,
  errors: 0,
  expires: { volatile: 0 },
  bytes: { read: 43210630, readThrough: 4951 } }
````

Specifically, cached results will be returned for all endpoints less than 800ms old. Anything more out-of-date than that will cause a device request. Currently the `device` and `network` endpoints are considered static, and will be cached indefinitely.
	
## Test Coverage

Clone the repository and install dev dependencies with `git clone git://github.com/cscade/BCS.client.git && cd BCS.client && npm install`.

Test coverage includes;

* `Device`
* `Cache`

To run Device and Cache tests, run `make test TARGET_HOST=192.168.1.100 TARGET_PORT=80`, substituting your BCS device's ip and port.

## License 

(The MIT License)

Copyright (c) 2012 Carson Christian &lt;cc@seekerbeer.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

[ecc]: http://www.embeddedcontrolconcepts.com/ "Welcome to Embedded Control Concepts"
[api-docs]: http://www.embeddedcc.com/api-docs/ "Open Interface API - ECC Learning Center"

History.md
## 1.0.1 - 205-11-10

* use setImmediate() instead of process.nextTick() for compatibility with modern node.js versions

## 1.0.0 - 2015-11-10

* full rewrite to support 4.x series BCS firmware

## 0.1.3 - 2012-11-27

* add full test suite for device, cache, and dictionary
* separate & flatten BCS-460/462 dictionaries, device detection auto-selects
* fix incorrect responses for temp setpoints on BCS-462

## 0.1.2 - 2012-11-20

* add device communication timeout
* improved error handling on initialization of a new device

## 0.1.1 - 2012-11-08

* rework dictionary to ensure full coverage on 460/462 devices
* improved docs

## 0.1.0 - 2012-11-08

* initial functioning release

test.js
// 
//  device_test.js
//  BCS.client
//  
//  Copyright (c) 2012 Carson S. Christian <cc@seekerbeer.com>
//  The MIT License
// 

var Cache = require('./lib/cache').Cache,
	Device = require('./lib/device').Device;

var tests = [
	function () {
		var cache = new Cache({ host: '192.168.42.2' });
		var structures = {
				'sysname.dat': 'static',
				'ipaddr.dat': 'static',
				'ultemp.dat': 'volatile',
				'ulsysio.dat': 'volatile',
				'ucsysio.dat': 'volatile',
				'enetlog.dat': 'volatile',
			}, done = 0;
		
		console.log('*** START: Test read-through on all structures');
		Object.keys(structures).forEach(function (struct) {
			var iterations = 10000;
	
			structures[struct] = { responses: 0 };
			for (var i = 1; i <= iterations; i++) {
				cache.get(struct, function (e, data) {
					structures[struct].responses++;
					if (e) console.log(struct, e);
					if (iterations === structures[struct].responses) {
						done++;
						console.log('done', struct);
					}
					if (done === Object.keys(structures).length) {
						console.log(structures);
						console.log(cache.stats);
						console.log('*** COMPLETE: Test read-through on all structures');
						(tests.shift() || function () {})();
					}
				});
			}
		});
	},
	function () {
		var device;
		
		console.log('*** START: Device initialization test');
		device = new Device('192.168.42.2', 80, function (e, state) {
			console.log(e, state);
			console.log('*** COMPLETE: Device initialization test');
		});
	},
];

tests.shift()();

test/cache.js
// 
//  test/cache.js
//  BCS.client
//  
//  Created by Carson S. Christian on 2012-11-26.
// 

var target = {
	host: process.env.TARGET_HOST,
	port: process.env.TARGET_PORT || 80
};

var should = require('should');

var Cache = require('../lib/cache').Cache;

describe('Cache', function () {
	var cache = new Cache({ host: target.host, port: target.port });
	
	it('should expose its host and port', function () {
		should.exist(cache.location);
		cache.location.host.should.equal(target.host);
		cache.location.port.should.equal(target.port);
	});
	describe('multiple endpoint requests', function () {
		this.timeout(10000);
		beforeEach(function () {
			cache = new Cache({ host: target.host, port: target.port });
		});
		afterEach(function () {
			cache.stats.errors.should.equal(0);
		});
		// static (non-volatile) endpoints
		['device', 'network'].forEach(function (endpoint) {
			describe('to ' + endpoint, function () {
				it('should not cause multiple device requests', function (done) {
					cache.get(endpoint, function (e, data) {
						if (e) throw e;
						setTimeout(function () {
							cache.get(endpoint, function (e, data) {
								if (e) throw e;
								cache.stats.hits.should.equal(1);
								cache.stats.misses.should.equal(1);
								cache.stats.expires.volatile.should.equal(0);
								done();
							});
						}, 2000);
					});
				});
			});
		});
		// volatile endpoints
		['alarm', 'din', 'output', 'state', 'system', 'temp'].forEach(function (endpoint) {
			describe('to ' + endpoint, function () {
				it('should cause multiple device requests', function (done) {
					cache.get(endpoint, function (e, data) {
						if (e) throw e;
						setTimeout(function () {
							cache.get(endpoint, function (e, data) {
								if (e) throw e;
								cache.stats.hits.should.equal(0);
								cache.stats.misses.should.equal(2);
								cache.stats.expires.volatile.should.equal(1);
								done();
							});
						}, 1000);
					});
				});
			});
		});
	});
});


lib/dictionary.js
// 
//  dictionary.js
//  BCS.client
//  
//  Copyright (c) 2012 Carson S. Christian <cc@seekerbeer.com>
//  The MIT License
// 

var extend = require('xtend');

// base dictionary for 460
var dictionary = {
	'bcs-460': {
		ps0: {
			name: {
				target: 'sysname.dat',
				limits: [1],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 0
			},
			state: {
				target: 'ultemp.dat',
				limits: [19],
				format: 'number',
				encoding: 'process-state',
				position: 0
			},
			state0: {
				target: 'sysname.dat',
				limits: [5],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [6],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [7],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [8],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [9],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [10],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [11],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [12],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [47],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [48],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [49],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [50],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [67],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [68],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [69],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [70],
				format: 'string'
			}
		},
		ps1: {
			name: {
				target: 'sysname.dat',
				limits: [2],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 1
			},
			state: {
				target: 'ultemp.dat',
				limits: [19],
				format: 'number',
				encoding: 'process-state',
				position: 1
			},
			state0: {
				target: 'sysname.dat',
				limits: [13],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [14],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [15],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [16],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [17],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [18],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [19],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [20],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [51],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [52],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [53],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [54],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [71],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [72],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [73],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [74],
				format: 'string'
			}
		},
		ps2: {
			name: {
				target: 'sysname.dat',
				limits: [3],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 2
			},
			state: {
				target: 'ultemp.dat',
				limits: [19],
				format: 'number',
				encoding: 'process-state',
				position: 2
			},
			state0: {
				target: 'sysname.dat',
				limits: [21],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [22],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [23],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [24],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [25],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [26],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [27],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [28],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [55],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [56],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [57],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [58],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [75],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [76],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [77],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [78],
				format: 'string'
			}
		},
		ps3: {
			name: {
				target: 'sysname.dat',
				limits: [4],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 3
			},
			state: {
				target: 'ultemp.dat',
				limits: [19],
				format: 'number',
				encoding: 'process-state',
				position: 3
			},
			state0: {
				target: 'sysname.dat',
				limits: [29],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [30],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [31],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [32],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [33],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [34],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [35],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [36],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [59],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [60],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [61],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [62],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [79],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [80],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [81],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [82],
				format: 'string'
			}
		},
		ps4: {
			name: {
				target: 'sysname.dat',
				limits: [105],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 4
			},
			state: {
				target: 'ultemp.dat',
				limits: [20],
				format: 'number',
				encoding: 'process-state',
				position: 0
			},
			state0: {
				target: 'sysname.dat',
				limits: [109],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [110],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [111],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [112],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [113],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [114],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [115],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [116],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [141],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [142],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [143],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [144],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [157],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [158],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [159],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [160],
				format: 'string'
			}
		},
		ps5: {
			name: {
				target: 'sysname.dat',
				limits: [106],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 5
			},
			state: {
				target: 'ultemp.dat',
				limits: [20],
				format: 'number',
				encoding: 'process-state',
				position: 1
			},
			state0: {
				target: 'sysname.dat',
				limits: [117],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [118],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [119],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [120],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [121],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [122],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [123],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [124],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [145],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [146],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [147],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [148],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [161],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [162],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [163],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [164],
				format: 'string'
			}
		},
		ps6: {
			name: {
				target: 'sysname.dat',
				limits: [107],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 6
			},
			state: {
				target: 'ultemp.dat',
				limits: [20],
				format: 'number',
				encoding: 'process-state',
				position: 2
			},
			state0: {
				target: 'sysname.dat',
				limits: [125],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [126],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [127],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [128],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [129],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [130],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [131],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [132],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [149],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [150],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [151],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [152],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [165],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [166],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [167],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [168],
				format: 'string'
			}
		},
		ps7: {
			name: {
				target: 'sysname.dat',
				limits: [108],
				format: 'string'
			},
			status: {
				target: 'ultemp.dat',
				limits: [0],
				format: 'boolean',
				encoding: 'binary',
				position: 7
			},
			state: {
				target: 'ultemp.dat',
				limits: [20],
				format: 'number',
				encoding: 'process-state',
				position: 3
			},
			state0: {
				target: 'sysname.dat',
				limits: [133],
				format: 'string'
			},
			state1: {
				target: 'sysname.dat',
				limits: [134],
				format: 'string'
			},
			state2: {
				target: 'sysname.dat',
				limits: [135],
				format: 'string'
			},
			state3: {
				target: 'sysname.dat',
				limits: [136],
				format: 'string'
			},
			state4: {
				target: 'sysname.dat',
				limits: [137],
				format: 'string'
			},
			state5: {
				target: 'sysname.dat',
				limits: [138],
				format: 'string'
			},
			state6: {
				target: 'sysname.dat',
				limits: [139],
				format: 'string'
			},
			state7: {
				target: 'sysname.dat',
				limits: [140],
				format: 'string'
			},
			win0: {
				target: 'sysname.dat',
				limits: [153],
				format: 'string'
			},
			win1: {
				target: 'sysname.dat',
				limits: [154],
				format: 'string'
			},
			win2: {
				target: 'sysname.dat',
				limits: [155],
				format: 'string'
			},
			win3: {
				target: 'sysname.dat',
				limits: [156],
				format: 'string'
			},
			timer0: {
				target: 'sysname.dat',
				limits: [169],
				format: 'string'
			},
			timer1: {
				target: 'sysname.dat',
				limits: [170],
				format: 'string'
			},
			timer2: {
				target: 'sysname.dat',
				limits: [171],
				format: 'string'
			},
			timer3: {
				target: 'sysname.dat',
				limits: [172],
				format: 'string'
			}
		},
		temp: {
			name0: {
				target: 'sysname.dat',
				limits: [63],
				format: 'string'
			},
			name1: {
				target: 'sysname.dat',
				limits: [64],
				format: 'string'
			},
			name2: {
				target: 'sysname.dat',
				limits: [65],
				format: 'string'
			},
			name3: {
				target: 'sysname.dat',
				limits: [66],
				format: 'string'
			},
			value0: {
				target: 'ultemp.dat',
				limits: [3],
				format: 'number',
				encoding: 'temperature-10'
			},
			value1: {
				target: 'ultemp.dat',
				limits: [4],
				format: 'number',
				encoding: 'temperature-10'
			},
			value2: {
				target: 'ultemp.dat',
				limits: [5],
				format: 'number',
				encoding: 'temperature-10'
			},
			value3: {
				target: 'ultemp.dat',
				limits: [6],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint0: {
				target: 'ultemp.dat',
				limits: [11],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint1: {
				target: 'ultemp.dat',
				limits: [12],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint2: {
				target: 'ultemp.dat',
				limits: [13],
				format: 'number',
				encoding: 'temperature-10'
			},
			setpoint3: {
				target: 'ultemp.dat',
				limits: [14],
				format: 'number',
				encoding: 'temperature-10'
			}
		},
		input: {
			name0: {
				target: 'sysname.dat',
				limits: [43],
				format: 'string'
			},
			name1: {
				target: 'sysname.dat',
				limits: [44],
				format: 'string'
			},
			name2: {
				target: 'sysname.dat',
				limits: [45],
				format: 'string'
			},
			name3: {
				target: 'sysname.dat',
				limits: [46],
				format: 'string'
			},
			status0: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 0
			},
			status1: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 1
			},
			status2: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 2
			},
			status3: {
				target: 'ultemp.dat',
				limits: [2],
				format: 'boolean',
				encoding: 'binary',
				position: 3
			}
		},
		output: {
			name0: {
				target: 'sysname.dat',
				limits: [37],
				format: 'string'
			},
			name1: {
				target: 'sysname.dat',
				limits: [38],
				format: 'string'
			},
			name2: {
				target: 'sysname.dat',
				limits: [39],
				format: 'string'
			},
			name3: {
				target: 'sysname.dat',
				limits: [40],
				format: 'string'
			},
			name4: {
				target: 'sysname.dat',
				limits: [41],
				format: 'string'
			},
			name5: {
				target: 'sysname.dat',
				limits: [42],
				format: 'string'
			},
			status0: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 0
			},
			status1: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 1
			},
			status2: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 2
			},
			status3: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 3
			},
			status4: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 4
			},
			status5: {
				target: 'ultemp.dat',
				limits: [1],
				format: 'boolean',
				encoding: 'binary',
				position: 5
			}
		},
		network: {
			staticAddress: {
				target: 'ipaddr.dat',
				limits: [0,3],
				format: 'string',
				join: '.'
			},
			subnetMask: {
				target: 'ipaddr.dat',
				limits: [4,7],
				format: 'string',
				join: '.'
			},
			gateway: {
				target: 'ipaddr.dat',
				limits: [8,11],
				format: 'string',
				join: '.'
			},
			MAC: {
				target: 'ipaddr.dat',
				limits: [12,17],
				format: 'string',
				encoding: 'hex',
				join: ':'
			},
			currentIP: {
				target: 'ipaddr.dat',
				limits: [18,21],
				format: 'string',
				join: '.'
			},
			DHCPEnabled: {
				target: 'ipaddr.dat',
				limits: [22],
				format: 'boolean'
			},
			packetsSent: {
				target: 'enetlog.dat',
				limits: [0],
				format: 'number'
			},
			packetsReceived: {
				target: 'enetlog.dat',
				limits: [2],
				format: 'number'
			}
		},
		system: {
			model: {
				target: 'ucsysio.dat',
				limits: [35],
				format: 'string',
				encoding: 'multiple-choice',
				choices: {
					0: 'BCS-460',
					2: 'BCS-462'
				}
			},
			firmware: {
				target: 'sysname.dat',
				limits: [0],
				format: 'string'
			}
		}
	}
};

// copy 460 dictionary to 462
dictionary['bcs-462'] = extend({}, dictionary['bcs-460']);
// extend with extra 462 temp probes
extend(dictionary['bcs-462'].temp, {
	name4: {
		target: 'sysname.dat',
		limits: [85],
		format: 'string'
	},
	name5: {
		target: 'sysname.dat',
		limits: [86],
		format: 'string'
	},
	name6: {
		target: 'sysname.dat',
		limits: [87],
		format: 'string'
	},
	name7: {
		target: 'sysname.dat',
		limits: [88],
		format: 'string'
	},
	value4: {
		target: 'ultemp.dat',
		limits: [7],
		format: 'number',
		encoding: 'temperature-10'
	},
	value5: {
		target: 'ultemp.dat',
		limits: [8],
		format: 'number',
		encoding: 'temperature-10'
	},
	value6: {
		target: 'ultemp.dat',
		limits: [9],
		format: 'number',
		encoding: 'temperature-10'
	},
	value7: {
		target: 'ultemp.dat',
		limits: [10],
		format: 'number',
		encoding: 'temperature-10'
	},
	setpoint4: {
		target: 'ultemp.dat',
		limits: [7],
		format: 'number',
		encoding: 'temperature-10'
	},
	setpoint5: {
		target: 'ultemp.dat',
		limits: [8],
		format: 'number',
		encoding: 'temperature-10'
	},
	setpoint6: {
		target: 'ultemp.dat',
		limits: [9],
		format: 'number',
		encoding: 'temperature-10'
	},
	setpoint7: {
		target: 'ultemp.dat',
		limits: [10],
		format: 'number',
		encoding: 'temperature-10'
	}
});
// extend with extend 462 inputs
extend(dictionary['bcs-462'].input, {
	name4: {
		target: 'sysname.dat',
		limits: [89],
		format: 'string'
	},
	name5: {
		target: 'sysname.dat',
		limits: [90],
		format: 'string'
	},
	name6: {
		target: 'sysname.dat',
		limits: [91],
		format: 'string'
	},
	name7: {
		target: 'sysname.dat',
		limits: [92],
		format: 'string'
	},
	status4: {
		target: 'ultemp.dat',
		limits: [2],
		format: 'boolean',
		encoding: 'binary',
		position: 4
	},
	status5: {
		target: 'ultemp.dat',
		limits: [2],
		format: 'boolean',
		encoding: 'binary',
		position: 5
	},
	status6: {
		target: 'ultemp.dat',
		limits: [2],
		format: 'boolean',
		encoding: 'binary',
		position: 6
	},
	status7: {
		target: 'ultemp.dat',
		limits: [2],
		format: 'boolean',
		encoding: 'binary',
		position: 7
	}
});
// extend with extra 462 outputs
extend(dictionary['bcs-462'].output, {
	name6: {
		target: 'sysname.dat',
		limits: [93],
		format: 'string'
	},
	name7: {
		target: 'sysname.dat',
		limits: [94],
		format: 'string'
	},
	name8: {
		target: 'sysname.dat',
		limits: [95],
		format: 'string'
	},
	name9: {
		target: 'sysname.dat',
		limits: [96],
		format: 'string'
	},
	name10: {
		target: 'sysname.dat',
		limits: [97],
		format: 'string'
	},
	name11: {
		target: 'sysname.dat',
		limits: [98],
		format: 'string'
	},
	name12: {
		target: 'sysname.dat',
		limits: [99],
		format: 'string'
	},
	name13: {
		target: 'sysname.dat',
		limits: [100],
		format: 'string'
	},
	name14: {
		target: 'sysname.dat',
		limits: [101],
		format: 'string'
	},
	name15: {
		target: 'sysname.dat',
		limits: [102],
		format: 'string'
	},
	name16: {
		target: 'sysname.dat',
		limits: [103],
		format: 'string'
	},
	name17: {
		target: 'sysname.dat',
		limits: [104],
		format: 'string'
	},
	status6: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 6
	},
	status7: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 7
	},
	status8: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 8
	},
	status9: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 9
	},
	status10: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 10
	},
	status11: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 11
	},
	status12: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 12
	},
	status13: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 13
	},
	status14: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 14
	},
	status15: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 15
	},
	status16: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 16
	},
	status17: {
		target: 'ultemp.dat',
		limits: [1],
		format: 'boolean',
		encoding: 'binary',
		position: 17
	}
});

exports.dictionary = dictionary;

